spamc.go

```diff
@@ -67,10 +68,148 @@ func (c *Client) send(
 	return conn, nil
 }
 
+// rearrangeMessageReader will rearrange the message reader to make sure SA will understand the message
+func (c Client) rearrangeMessageReader(message io.ReadSeeker, size int64) (io.Reader, int64, error) {
```

Comment:
I'm a bit confused what the purpose of this function is exactly; what needs to be "rearranged" exactly, and why? It looks like it's doing something with the message's email headers?

It would be good to add a `TestRearrangeMessageReader()` for this function, which expected input and output.

Reply:
This function basically reads the message headers as well as it can, it needs to do this to ensure that the message can be split into <headers>\r\n<body>\r\n both headers and body may be empty but those \r\n need to be there.  It will insert both these endings if:
1) headers end abruptly due to invalid syntax (like none ascii character, missing : key/value terminator, etc)
2) body does not end in \r\n

It will return a reader that always contains those \r\n and alters the size to if it inserted these, this would make the any reader content passed into this function something that SA understands without timing out on the read.

Reply:
>  headers end abruptly due to invalid syntax (like none ascii character, missing : key/value terminator, etc)

Is this a problem in the sense that SA will hang the connection? Because various software (both spam and non-spam) *does* send invalid emails unfortunately. I'm not sure if this "ad-hoc parsing" of that is a good idea to be honest.

by the way, the way I had in mind to add the trailing newline is something like this (untested, just to show the concept):

```
for {
	// Read 1K from the input.
	data, err := inputReader.Read(1024)

	if err != nil {
		// End of file, make sure the last chunck ended with \r\n
		if err == io.EOF {
			if !bytes.HasSuffix(data, "\r\n") {
				spamcWriter.Write("\r\n")
			}
             break
		}
	}

	// Write to SpamAssassin
	spamdWriter.Write(data)
}
```

Reply:
> would also need to increase the Content-length header, and this was already sent

Ah yes, of course >_<

I did some testing, and it seems that it *should* work without the new line; sith this message:

	[~]% hexdump -C test-msg
	00000000  66 6f 6f                                          |foo|
	00000003

The "official" spamc client seems to work okay:

	[~]% spamc --check < test-msg
	7.9/5.0

It seems to send (`nc -l 783 | hexdump -C`):

```
00000000  43 48 45 43 4b 20 53 50  41 4d 43 2f 31 2e 35 0d  |CHECK SPAMC/1.5.|
00000010  0a 55 73 65 72 3a 20 6d  61 72 74 69 6e 0d 0a 43  |.User: martin..C|
00000020  6f 6e 74 65 6e 74 2d 6c  65 6e 67 74 68 3a 20 33  |ontent-length: 3|
00000030  0d 0a 0d 0a 66 6f 6f                              |....foo|
00000037
```

without any trailing newline.

So ... I don't know why we need to send this newline. We should probably fix that somewhere else.

I'll see if I can fix that in a new PR; then we can remove all of this stuff :-)

Reply:
I took a look at the spamc C code, they are closing the connection for writing after sending the last char (also with no newline termination).  

As a test I removed the rearrangeMessageReader call and added the following after the io.Copy block.

```Go
if tc, ok := conn.(*net.TCPConn); ok {
	return tc.CloseWrite()
}
```

The connection did not stall, and an immediate response was received,  we are on the right track here good debugging btw ;)